services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java

```diff
@@ -0,0 +1,254 @@
+                String rangeId = UUID.randomUUID().toString();
+                NetworkVlanRangeRequest request = new NetworkVlanRangeRequest(rangeId, segmentId, networkType, NetworkType.VLAN_FIRST_KEY, NetworkType.VLAN_LAST_KEY);
+                this.vlanRangeRepository.createRange(request);
+            }
+
+            String rangeId = "";
+            for (Map.Entry<String, NetworkVlanRange> entry : map.entrySet()) {
+                rangeId = entry.getKey();
+            }
+
+            Long key = this.vlanRangeRepository.allocateVlanKey(rangeId);
+
+            //vlan.setMtu(mtu);
+            vlan.setSegmentId(segmentId);
+            vlan.setVlanId(vlanId);
+            vlan.setKey(key);
+
+            this.vlanRepository.addItem(vlan);
+            logger.info("Allocate vlan key success, key: " + key);
+            return key;
+        } catch (Exception e) {
+            this.vlanRepository.deleteItem(vlanId);
+            logger.info("Allocate vlan key failed");
+            throw new DatabasePersistenceException(e.getMessage());
+        }
+
+    }
+
+    @Override
+    public Long addVxlanEntity(String segmentId) throws DatabasePersistenceException {
+        try {
+            String vxlanId = UUID.randomUUID().toString();
+            NetworkVxlanType vxlan = new NetworkVxlanType();
+            //vxlan.setMtu(mtu);
+            vxlan.setSegmentId(segmentId);
+            vxlan.setVxlanId(vxlanId);
+
+            this.vxlanRepository.addItem(vxlan);
+            return null;
+        } catch (Exception e) {
+            throw new DatabasePersistenceException(e.getMessage());
+        }
+    }
+
+    @Override
+    public Long addGreEntity(String segmentId) throws DatabasePersistenceException {
+        try {
+            String greId = UUID.randomUUID().toString();
+            NetworkVGREType gre = new NetworkVGREType();
+            gre.setSegmentId(segmentId);
+            gre.setGreId(greId);
+
+            this.greRepository.addItem(gre);
+            return null;
+        } catch (Exception e) {
+            throw new DatabasePersistenceException(e.getMessage());
+        }
+    }
+
+    @Override
+    public void releaseVlanEntity(String vlanId, Long key) throws DatabasePersistenceException {
+        try {
+            Map<String, NetworkVlanRange> map = this.vlanRangeRepository.findAllItems();
+            if (map.size() == 0) {
+                // TO DO: throw exception
+                return;
+            }
+
+            String rangeId = "";
+            for (Map.Entry<String, NetworkVlanRange> entry : map.entrySet()) {
+                rangeId = entry.getKey();
+            }
+
+            this.vlanRangeRepository.releaseVlanKey(rangeId, key);
+            this.vlanRepository.deleteItem(vlanId);
+        } catch (Exception e) {
+            throw new DatabasePersistenceException(e.getMessage());
+        }
+    }
+
+    @Override
+    public VlanKeyRequest allocateVlan(VlanKeyRequest request) throws Exception {
```

Comment:
Response should be "VlanKeyResponse", right?

Reply:
Yeah ~ althought VlanKeyResponse is same as VlanKeyRequest here